spec/integration/forking_spec.rb

```diff
@@ -0,0 +1,247 @@
+# frozen_string_literal: true
+
+require "racecar/cli"
+
+RSpec.describe "forking", type: :integration do
+  let!(:racecar_cli) { Racecar::Cli.new(["ForkingConsumer"]) }
+  let(:input_topic) { generate_input_topic_name }
+  let(:output_topic) { generate_output_topic_name }
+  let(:group_id) { generate_group_id }
+
+  let(:input_messages) do
+    total_messages.times.map { |n| { payload: "message-#{n}", partition: n % topic_partitions } }
+  end
+
+  let(:total_messages) { messages_per_topic * topic_partitions }
+  let(:topic_partitions) { 4 }
+  let(:messages_per_topic) { 10 }
+  let(:forks) { 2 }
+  let(:consumer_class) { ForkingConsumer ||= echo_consumer_class }
+
+  before do
+    create_topic(topic: input_topic, partitions: topic_partitions)
+    create_topic(topic: output_topic, partitions: topic_partitions)
+
+    consumer_class.subscribes_to(input_topic)
+    consumer_class.output_topic = output_topic
+    consumer_class.group_id = group_id
+    consumer_class.pipe_to_test = consumer_message_pipe
+
+    Racecar.config.forks = forks
+  end
+
+  after do |test|
+    Object.send(:remove_const, :ForkingConsumer) if defined?(ForkingConsumer)
+  end
+
+  it "each fork consumes messages in parallel" do
+    start_racecar
+
+    wait_for_assignments(forks)
+    publish_messages
+    wait_for_messages
+
+    expect_equal_distribution_of_message_processing
+    expect_processing_times_to_mostly_overlap
+  end
+
+  context "when the supervisor process receives TERM" do
+    let(:messages_per_topic) { 1 }
+
+    it "terminates the worker processes" do
+      start_racecar
+
+      wait_for_assignments(forks)
+      publish_messages
+      wait_for_messages
+
+      Process.kill("TERM", supervisor_pid)
+      Process.wait(supervisor_pid)
+
+      expect_processes_to_have_exited(worker_pids)
+    end
+  end
+
+  context "when the supervisor process is killed (SIGKILL)" do
+    let(:messages_per_topic) { 1 }
+
+    it "terminates the worker processes" do
+      start_racecar
+
+      wait_for_assignments(forks)
+      publish_messages
+      wait_for_messages
+
+      Process.kill("KILL", supervisor_pid)
+      Process.waitpid(supervisor_pid)
+
+      expect_processes_to_have_exited(worker_pids)
+    end
+  end
+
+  context "when a worker process exits" do
+    let(:messages_per_topic) { 1 }
+
+    it "terminates all other processes gracefully" do
+      start_racecar
+
+      wait_for_assignments(forks)
+      publish_messages
+      wait_for_messages
+
+      Process.kill("KILL", worker_pids[0])
+      Process.waitpid(supervisor_pid)
+
+      expect_processes_to_have_exited(worker_pids)
```

Comment:
If we want to k8s to spin up a new replica shouldn't this also kill the `supervisor_pid`?